cranelift/codegen/src/machinst/abi.rs

```diff
@@ -304,6 +305,46 @@ pub trait IsaFlags: Clone {
     }
 }
 
+/// Used as an out-parameter to accumulate a sequence of `ABIArg`s in
+/// `ABIMachineSpec::compute_arg_locs`. Wraps the shared allocation for all
+/// `ABIArg`s in `SigSet` and exposes just the args for the current
+/// `compute_arg_locs` call.
+pub struct ArgsAccumulator<'a> {
+    sig_set_abi_args: &'a mut Vec<ABIArg>,
+    start: usize,
+}
+
+impl<'a> ArgsAccumulator<'a> {
+    fn new(sig_set_abi_args: &'a mut Vec<ABIArg>) -> Self {
+        let start = sig_set_abi_args.len();
+        ArgsAccumulator {
+            sig_set_abi_args,
+            start,
+        }
+    }
+
+    #[inline]
+    pub fn push(&mut self, arg: ABIArg) {
+        self.sig_set_abi_args.push(arg)
+    }
+}
+
+impl std::ops::Deref for ArgsAccumulator<'_> {
```

Comment:
I may be in the minority opinion on this but I have a slight preference against too much magic `Deref`: it took me a bit too long to figure out how `args.len()` was working (answer: `Deref` lets this become a slice and then `.len()` works on that). If so for me then doubly so for someone new to our codebase trying to learn how things work. I sort of expect smart pointer types to have `Deref` but I don't expect that for an "accumulator"!

Could we have a `.args()` and `.args_mut()` or similar to make it clear what's going on instead?

Reply:
How strongly do you feel about this? Uses will go from `args.len()` to `args.args().len()` which I don't feel is an improvement. `rust-analyzer` will happily jump to `[T]::len` when you do jump-to-definition on `args.len()`, so I feel like it isn't hidden. I can definitely highlight the deref impls in the doc comment for `ArgsAccumulator` as well.

Reply:
> I sort of expect smart pointer types to have `Deref`

I feel this is exactly as much of a smart pointer as `Vec` itself is, so I'd argue that if you're okay with `Vec` dereferencing to a slice, than this should be okay too.

> but I don't expect that for an "accumulator"!

Happy to bikeshed the name as well, I put approximately zero thought into it.

Reply:
Yeah, I don't feel too strongly about it, I just have a general discomfort with implicit magic: it imposes a cost on the reader to maintain some mental state when reading code. The most frustrating codebases for me to dive into are those where there is a lot of magic dispatch and implicit connection; I don't like to weave a maze that will confuse others like that.

(Also, I'd prefer not to have a codebase that requires `rust-analyzer` to comprehend well; not everyone has it installed, it can be frustrating to set up, it's fairly resource-intensive, etc.)

I actually didn't realize `Vec` used `Deref` to "inherit" all the methods on slices; that's good to know. I do feel that's a slightly different case though, as it's part of the standard library and part of the shared base of knowledge most Rust programmers will have; here we are defining a custom thing.

If we're using just a few methods from the slice API (is it just `.len()` that uses this in practice?) I'd prefer to hand-write the 3-line `fn len(&self) { ... }` that proxies to the inner slice. If that glue is more than a few methods then sure, let's have the `Deref` and make sure we make a note somewhere (in the main doc-comment for `ArgsAccumulator` maybe) that it auto-derefs to the slice.

Reply:
Some of the backends (s390x iirc) will also index into the slice and reassign args within slots. I don't feel super strongly so I'll just add an `args` method.